### ISSUE TITLE
Server monitoring

### ISSUE THREAD

**Issue by Unknwon** [gogits/gogs#4141]
Prometheus + Grafana

- [ ] Runtime metrics
- [ ] HTTP response
- [ ] Database response
- [ ] Git SSH/HTTP response

**Comment by appleboy** [comment]
Nice ... 👍

**Comment by otbutz** [comment]
Would anybody be interested in a minimal monit config?

**Comment by Unknwon** [comment]
@otbutz what do you mean?

**Comment by otbutz** [comment]
I'm talking about a configuration for the monit monitoring daemon (https://mmonit.com/monit).
On Debian based distributions monit already ships with some available configurations within `/etc/monit/conf-available` in order to make it easier for the user.
We could add a basic config for gogs and keep it in the repository like the systemd config file.

**Comment by Unknwon** [comment]
@otbutz Based on what I read from it's introduction:

>With all features needed for system monitoring and error recovery. It's like having a watchdog with a toolbox on your server

It's for server, not for application's internal monitoring.

**Comment by oliver006** [comment]
Monit is a great tool to make sure processes are staying up and similar tasks but serves a different use case than Prometheus which is for exposing, scraping and monitoring process metrics.

**Comment by Unknwon** [comment]
@oliver006 OK... maybe my initial post is confusing... this thread is about internal monitoring of Gogs... in code/request level, not application lifetime, etc.

**Comment by oliver006** [comment]
Exactly. Maybe my last post wasn't clear but I think Prometheus would be perfect for that.

**Comment by otbutz** [comment]
Why not both?

**Comment by ogarcia** [comment]
I think that the first step is make a statistics api access (that returns a json for example), because every person use his own monitoring system.

For example, in my case I'm using [InfluxDB][1]. If I'll have a statistics api I can build a fetcher to send these statistics to my InfluxDB.

[1]:https://docs.influxdata.com/influxdb/v1.2/

**Comment by oliver006** [comment]
I'd think that using e.g. Prometheus would save you a lot of work cause you wouldn't have to reinvent the wheel and implement a statistics API but could use existing libraries that provide Gauges, Couters, Summaries, etc.
 If you start with a json stats API you'd still need to build a lot of the plumbing yourself like the "fetcher" you mention.

Using Prometheus instrumenting would also allow you to scrape/export the statistics data to e.g. InfluxDB for mid/long-term storage and provide the benefits of the entire Prometheus eco system (like the alertmanager).

**Comment by ogarcia** [comment]
@oliver006 I don't say that cannot implement Prometheus, but take note that there are more systems as Whisper (Graphite), InfluxDB, Kairos, Riak, Elasticsearch... and every person may use any of them. If you ONLY implements Prometheus support you are wasting a lot of systems.

IMHO, the first step is have a statistics API and then implement the different connectors. Either as part of Gogs or as third part.

**Comment by oliver006** [comment]
Exactly!
Let Prometheus take care of the metrics accounting, scraping, etc and then you can decide whatever timeseries DB you'd prefer as a storage/querying backend.

**Comment by louzadod** [comment]
Prometheus will do the job.
+1 on this

**Comment by tomkukral** [comment]
I'd love to have `/metrics` expoted by gogs.

**Comment by d-riecken** [comment]
Does anyone know what's the status on this feature request?

**Comment by Unknwon** [comment]
@d-riecken **Runtime metrics** should be easy and out of box, I'll priority this part.

**Comment by lexfrei** [comment]
Is there any news?

**Comment by Unknwon** [comment]
@lexfrei basic Go runtime metrics (provided by Prometheus) will be supported within next next release.

**Comment by Unknwon** [comment]
Go runtime metrics provided by Prometheus is now added via path `/-/metrics`, it is enabled by default and you have option to enable HTTP Basic Authentication.

https://github.com/gogs/gogs/blob/7a7e07a57c232b11bdaf713011e574420fb280e4/conf/app.ini#L467-L471

Example config in `prometheus.yml`:

```yml
scrape_configs:
  - job_name: 'gogs'
    metrics_path: "/-/metrics"
    basic_auth:
      username: 123
      password: 456
    static_configs:
      - targets: ['localhost:3000']
        labels:
          "instance": "Demo"
```

**Comment by seboss666** [comment]
Hello,
I'm running gogs in docker swarm (0.11.68), I don't find how to use the /-/metrics URL in Grafana. I have to admit I'm pretty noobish on Prometheus, am I missing something ?

**Comment by ogarcia** [comment]
@seboss666 you need a time series database (Prometheus, InfluxDB, Graphite, Elasticsearch, etc.) and a collector agent (Telegraf, colletd, etc.) to pull data from `/-/metrics` periodically and read it with Grafana.

For me the most easy solution is use InfluxDB and Telegraf, but there are a lot of them.

**Comment by seboss666** [comment]
After adding some services in my cluster (the telegraf/influxdb/grafana stack), at least I have the data collected, no I'll have to look at how to make a proper dashboard for it, thanks :)

**Comment by Hadlock** [comment]
This worked for me:

prometheus config.yaml
```yaml
    - job_name: git-gogs
      scheme: https
      metrics_path: '/-/metrics'
      static_configs:
        - targets: ['git.example.net:443']
```

Using this default grafana chart:

https://grafana.com/dashboards/240

It is sparse but functional.

**Comment by Santosh-src** [comment]
When can we anticipate the availability of metrics beyond runtime data? We are in urgent need of metrics related to HTTP and Git SSH/HTTP usage to effectively monitor the repository's performance and its traffic. Kindly propose any alternative methods to fulfill this requirement of tracking the number of clones (via SSH/HTTP) occurring on a specific repository.
